### PR TITLE
[FIXED] Cluster: Handling of deleted channels

### LINKED ISSUE #1008: STAN restart failing with error "unable to restore messages, can't start until leader is available"
I've 3 node nats streaming cluster running and was running fine. Due to some issues, node goes down. Once my nodes have recovered I've tried restarting nats streaming server in all the three nodes but it never comes up.

logs -
[node_1.log](https://github.com/nats-io/nats-streaming-server/files/4057405/node_1.log)
[node_3.log](https://github.com/nats-io/nats-streaming-server/files/4057403/node_3.log)
[node_2.log](https://github.com/nats-io/nats-streaming-server/files/4057404/node_2.log)

config files -
[1_conf.txt](https://github.com/nats-io/nats-streaming-server/files/4057416/1_conf.txt)
[2_conf.txt](https://github.com/nats-io/nats-streaming-server/files/4057417/2_conf.txt)
[3_conf.txt](https://github.com/nats-io/nats-streaming-server/files/4057418/3_conf.txt)

(I had to redeploy the cluster with same config file)

### LINKED ISSUE #1010: NSS upgrade 0.9.2->0.16.2 breaks cluster and causes "unable to restore messages, can't start until leader is available"
Three-node cluster had been upgraded (all nodes at once due to proto change) and stopped functioning. I've seen another GH issue with "can't start until leader is available" couple of days ago, but option `sync: true` doesn't help much. Same upgrade procedure works well on a cluster with no data. Logs and config are pasted below, please suggest.
Much thanks in advance!

**BEFORE UPGRADE**
```
'tail /var/log/nats/nss.log'
env-nats-01:
    [14373] 2020/01/16 11:11:37.177927 [INF] STREAM: --------- Channels Limits --------
    [14373] 2020/01/16 11:11:37.177942 [INF] STREAM:   Subscriptions:          1000 *
    [14373] 2020/01/16 11:11:37.177960 [INF] STREAM:   Messages     :       1000000 *
    [14373] 2020/01/16 11:11:37.177979 [INF] STREAM:   Bytes        :     976.56 MB *
    [14373] 2020/01/16 11:11:37.177997 [INF] STREAM:   Age          :     unlimited *
    [14373] 2020/01/16 11:11:37.178010 [INF] STREAM:   Inactivity   :     unlimited *
    [14373] 2020/01/16 11:11:37.178051 [INF] STREAM: ----------------------------------
    [14373] 2020/01/16 11:11:37.472781 [INF] 172.21.0.5:35174 - rid:10 - Route connection created
    [14373] 2020/01/16 11:11:37.521777 [INF] 172.21.0.4:49120 - rid:11 - Route connection created
env-nats-03:
    [12234] 2020/01/16 11:11:37.425761 [INF] STREAM:   Messages     :       1000000 *
    [12234] 2020/01/16 11:11:37.425779 [INF] STREAM:   Bytes        :     976.56 MB *
    [12234] 2020/01/16 11:11:37.425798 [INF] STREAM:   Age          :     unlimited *
    [12234] 2020/01/16 11:11:37.425816 [INF] STREAM:   Inactivity   :     unlimited *
    [12234] 2020/01/16 11:11:37.425840 [INF] STREAM: ----------------------------------
    [12234] 2020/01/16 11:11:37.472874 [INF] 172.21.0.5:54086 - rid:10 - Route connection created
    [12234] 2020/01/16 11:11:37.523238 [INF] 172.21.0.3:6222 - rid:11 - Route connection created
    [12234] 2020/01/16 11:11:39.213455 [INF] STREAM: Deleting raft logs from 42199461 to 42199461
    [12234] 2020/01/16 11:11:39.213733 [INF] STREAM: Deletion took 279.6µs
env-nats-02:
    [9961] 2020/01/16 11:11:37.882306 [INF] STREAM: --------- Channels Limits --------
    [9961] 2020/01/16 11:11:37.882324 [INF] STREAM:   Subscriptions:          1000 *
    [9961] 2020/01/16 11:11:37.882342 [INF] STREAM:   Messages     :       1000000 *
    [9961] 2020/01/16 11:11:37.882360 [INF] STREAM:   Bytes        :     976.56 MB *
    [9961] 2020/01/16 11:11:37.882378 [INF] STREAM:   Age          :     unlimited *
    [9961] 2020/01/16 11:11:37.882396 [INF] STREAM:   Inactivity   :     unlimited *
    [9961] 2020/01/16 11:11:37.882414 [INF] STREAM: ----------------------------------
    [9961] 2020/01/16 11:11:39.207654 [INF] STREAM: server became leader, performing leader promotion actions
    [9961] 2020/01/16 11:11:39.389969 [INF] STREAM: finished leader promotion actions
```

'curl -s localhost:8223/streaming/serverz | grep state'
```
env-nats-01:
      "state": "CLUSTERED",
env-nats-02:
      "state": "CLUSTERED",
env-nats-03:
      "state": "CLUSTERED",
```

'nats-streaming-server --version'
```
env-nats-03:
    nats-streaming-server version 0.9.2, nats-server version 1.0.7
env-nats-01:
    nats-streaming-server version 0.9.2, nats-server version 1.0.7
env-nats-02:
    nats-streaming-server version 0.9.2, nats-server version 1.0.7
```

**AFTER UPGRADE**
'nats-streaming-server --version'
```
env-nats-02:
    nats-streaming-server version 0.16.2, nats-server: v2.0.4
env-nats-01:
    nats-streaming-server version 0.16.2, nats-server: v2.0.4
env-nats-03:
    nats-streaming-server version 0.16.2, nats-server: v2.0.4
```

'curl -s localhost:8223/streaming/serverz | grep state'
```
env-nats-02:
      "state": "CLUSTERED",
env-nats-03:
	  curl: connection timeout
env-nats-01:
	  curl: connection timeout
```

'tail /var/log/nats/nss.log'
```
env-nats-01:
    [15532] 2020/01/16 11:19:16.503440 [INF] 172.21.0.5:35776 - rid:11 - Router connection closed
    [15532] 2020/01/16 11:19:18.083040 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [15532] 2020/01/16 11:19:19.619417 [INF] 172.21.0.4:6222 - rid:4 - Router connection closed
    [15532] 2020/01/16 11:19:19.763567 [INF] 172.21.0.4:49722 - rid:12 - Route connection created
    [15532] 2020/01/16 11:19:20.664472 [INF] 172.21.0.4:6222 - rid:13 - Route connection created
    [15532] 2020/01/16 11:19:20.664603 [INF] 172.21.0.4:6222 - rid:13 - Router connection closed
    [15532] 2020/01/16 11:19:21.086269 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [15532] 2020/01/16 11:19:24.088864 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [15532] 2020/01/16 11:19:27.056425 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [15532] 2020/01/16 11:19:30.062032 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
env-nats-03:
    [13385] 2020/01/16 11:19:20.093565 [INF] STREAM: Cluster Node ID : env-nats-03
    [13385] 2020/01/16 11:19:20.093586 [INF] STREAM: Cluster Log Path: env-nats-streaming/env-nats-03
    [13385] 2020/01/16 11:19:20.664173 [INF] 172.21.0.3:52446 - rid:10 - Route connection created
    [13385] 2020/01/16 11:19:20.664417 [INF] 172.21.0.3:52446 - rid:10 - Router connection closed
    [13385] 2020/01/16 11:19:20.668360 [INF] 172.21.0.5:54706 - rid:11 - Route connection created
    [13385] 2020/01/16 11:19:20.668519 [INF] 172.21.0.5:54706 - rid:11 - Router connection closed
    [13385] 2020/01/16 11:19:22.096915 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [13385] 2020/01/16 11:19:25.099436 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [13385] 2020/01/16 11:19:28.066863 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
    [13385] 2020/01/16 11:19:31.070344 [ERR] STREAM: channel "xxxxxxxx_failure" - unable to restore messages, can't start until leader is available
env-nats-02:
    [10906] 2020/01/16 11:18:50.164052 [INF] 172.21.0.4:6222 - rid:39 - Route connection created
    [10906] 2020/01/16 11:18:50.164644 [INF] 172.21.0.4:6222 - rid:39 - Router connection closed
    [10906] 2020/01/16 11:19:15.398990 [INF] 172.21.0.3:36426 - rid:36 - Router connection closed
    [10906] 2020/01/16 11:19:15.770327 [INF] 172.21.0.3:36468 - rid:40 - Route connection created
    [10906] 2020/01/16 11:19:16.502866 [INF] 172.21.0.3:6222 - rid:41 - Route connection created
    [10906] 2020/01/16 11:19:16.502866 [INF] 172.21.0.3:6222 - rid:41 - Router connection closed
    [10906] 2020/01/16 11:19:19.619301 [INF] 172.21.0.4:38584 - rid:38 - Router connection closed
    [10906] 2020/01/16 11:19:19.763688 [INF] 172.21.0.4:38628 - rid:42 - Route connection created
    [10906] 2020/01/16 11:19:20.668102 [INF] 172.21.0.4:6222 - rid:43 - Route connection created
    [10906] 2020/01/16 11:19:20.668381 [INF] 172.21.0.4:6222 - rid:43 - Router connection closed
```

**CONFIGS**
cat /etc/nats-streaming-server.conf
```
port: 4222
http_port: 8223
# pid file
pid_file: "/home/nats/run/nss.pid"

# logging options
debug:   false
trace:   false
logtime: true
log_file: "/var/log/nats/nss.log"

cluster {
  listen: 0.0.0.0:6222
  routes: [ "nats://env-nats-01:6222",   "nats://env-nats-02:6222",   "nats://env-nats-03:6222" ]
}

# NATS Streaming specific configuration
streaming {
  id: env-nats-streaming
  store: file
  dir: store
  cluster {
    node_id: "env-nats-01"
    peers: [ "env-nats-01",   "env-nats-02",   "env-nats-03" ]
    sync: true
  }
}
```

### PR DESCRIPTION
**Pull request by kozlovic**
A channel could be deleted and then later recreated (same name).
This is not handled properly in some cases such as restore from
snapshot.
This PR attempts to fix the problem by storing in the RAFT db
a UID when a channel is created. This UID is added to some of
the replicated operations.

Resolves #1008
Resolves #1010

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: server/clustering.go [hunk 1]
```diff
@@ -297,6 +297,20 @@ func (s *StanServer) createRaftNode(name string) (bool, error) {
 	if err != nil {
 		return false, err
 	}
+
+	// Get through the list of channels that we have recovered from streaming store
```
Comment (derekcollison) [inline server/clustering.go]: Get or Go?

File: server/clustering.go [hunk 2]
```diff
@@ -513,40 +533,42 @@ func (r *raftFSM) Apply(l *raft.Log) interface{} {
...
+			c.lSeqChecked = true
+		}
+		for _, msg = range op.PublishBatch.Messages {
+			if _, err = c.store.Msgs.Store(msg); err != nil {
+				goto FATAL_ERROR
 			}
 		}
 		return c.store.Msgs.Flush()
+	FATAL_ERROR:
+		panic(fmt.Errorf("failed to store replicated message %d on channel %s: %v",
```
Comment (derekcollison) [inline server/clustering.go]: No way to recover from this?

File: server/clustering.go [hunk 3]
```diff
@@ -572,9 +602,75 @@ func (r *raftFSM) Apply(l *raft.Log) interface{} {
...
+// This returns a channel object only if it is found with the proper ID.
+func (r *raftFSM) lookupChannel(name string, id uint64) *channel {
+	s := r.server
+	cs := s.channels
+	cs.RLock()
+	defer cs.RUnlock()
+
+	c := cs.channels[name]
+	// Consider no ID (either in channel or from param) to be a match.
+	if c != nil && (id == 0 || c.id == 0 || c.id == id) {
```
Comment (derekcollison) [inline server/clustering.go]: What does id == 0 signify?

File: server/server.go [hunk 4]
```diff
@@ -572,6 +584,14 @@ func (s *StanServer) subToSnapshotRestoreRequests() error {
 					return
 				}
 				sendingTheFirstAvail = msg != nil
+				if sendingTheFirstAvail && msg.Sequence < seq {
+					// it could be that a follower tries to restore from
```
Comment (derekcollison) [inline server/server.go]: Cap it

File: server/server.go [hunk 5]
```diff
@@ -572,6 +584,14 @@ func (s *StanServer) subToSnapshotRestoreRequests() error {
...
+					// snapshot of previous version of same channel name
```
Comment (derekcollison) [inline server/server.go]: "a snapshot"

Comment (derekcollison) [inline server/server.go]: of the same

File: server/snapshot.go [hunk 6]
```diff
@@ -380,10 +353,25 @@ func (r *raftFSM) restoreChannelsFromSnapshot(serverSnap *spb.RaftSnapshot, inNe
...
 	for _, sc := range serverSnap.Channels {
-		c, err := s.lookupOrCreateChannel(sc.Channel)
-		if err != nil {
-			return false, err
+		var c *channel
+		var err error
+		if inNewRaftCall {
+			// When starting the node and recovering from a local snapshot,
+			// ignore if the channel did not exist in our streaming store
+			// or if its UID does not match the one in the snapshot operation.
```
Comment (derekcollison) [inline server/snapshot.go]: You call it ID everywhere else.

File: server/snapshot.go [hunk 7]
```diff
@@ -380,10 +353,25 @@ func (r *raftFSM) restoreChannelsFromSnapshot(serverSnap *spb.RaftSnapshot, inNe
...
+			if c = r.lookupChannel(sc.Channel, sc.ChannelID); c == nil {
+				continue
+			}
+		} else {
+			// When restoring from a snapshot at runtime, this function will
+			// return the channel if it has the same ID, otherwise will create
+			// the channel (will first delete the old one if exits).
```
Comment (derekcollison) [inline server/snapshot.go]: if it exists

File: server/clustering.go (hunk 1 above)
Comment (kozlovic) [reply server/clustering.go]: fixed

File: server/clustering.go (hunk 2 above)
Comment (kozlovic) [reply server/clustering.go]: No. The RAFT log is committed at this point, if we fail to add to our message store we would take the risk of being out of sync and add later a higher sequence which will create gaps.

File: server/clustering.go (hunk 3 above)
Comment (kozlovic) [reply server/clustering.go]: This is for existing channels that were created before this fix.

File: server/server.go (hunk 4 above)
Comment (kozlovic) [reply server/server.go]: fixed

File: server/server.go (hunk 5 above)
Comment (kozlovic) [reply server/server.go]: fixed

Comment (kozlovic) [reply server/server.go]: fixed

File: server/snapshot.go (hunk 6 above)
Comment (kozlovic) [reply server/snapshot.go]: fixed

File: server/snapshot.go (hunk 7 above)
Comment (kozlovic) [reply server/snapshot.go]: fixed

File: server/clustering.go (hunk 2 above)
Comment (derekcollison) [reply server/clustering.go]: How would an operator restore a working system after this failure?

File: server/clustering.go (hunk 3 above)
Comment (derekcollison) [reply server/clustering.go]: Gotcha. Maybe add comment.

File: server/clustering.go (hunk 2 above)
Comment (kozlovic) [reply server/clustering.go]: The failure is due to not be able to write into the message store (disk full, storage failure)? So making some room or if the disk is toasted, the node would be restarted without any state and will be reconstructed from the running leader.

Comment (derekcollison) [reply server/clustering.go]: Thx

**Pull request merged**